### PR TITLE
Fix wrong denominator value when Period is financial year

### PR DESCRIPTION
**Pull request by luciano-fiandesio**

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/JdbcAnalyticsManager.java [hunk 1]
```diff
@@ -226,15 +228,60 @@ public void replaceDataPeriodsWithAggregationPeriods( Map<String, Object> dataVa
...
+
+    /**
+     * Calculates the factor for the weighted average
+     *
+     * @param periodType a {@link PeriodType}
+     * @return a value based on the passed period type
+     */
+    private int calculateWeightFactor(PeriodType periodType )
+    {
+
```
Comment (jimgrace) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/JdbcAnalyticsManager.java]: Remove blank line

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/JdbcAnalyticsManager.java [hunk 2]
```diff
@@ -226,15 +228,60 @@ public void replaceDataPeriodsWithAggregationPeriods( Map<String, Object> dataVa
...
+        return Precision.round( (year1Value * ((double) (12 - weight) / 12)) + (year2Value * ((double) weight / 12)), 2 );
+    }
+
+    /**
+     * Calculates the factor for the weighted average
+     *
+     * @param periodType a {@link PeriodType}
+     * @return a value based on the passed period type
+     */
+    private int calculateWeightFactor(PeriodType periodType )
```
Comment (jimgrace) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/JdbcAnalyticsManager.java]: Add space after (

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/JdbcAnalyticsManager.java [hunk 3]
```diff
@@ -226,15 +228,60 @@ public void replaceDataPeriodsWithAggregationPeriods( Map<String, Object> dataVa
...
+    private boolean requiresWeightedAvg( Period period  )
+    {
+        return period.getPeriodType().isFinancialYear();
+    }
+
+    private Double calculateWeightedAvg( Double year1Value, Double year2Value, PeriodType periodType )
+    {
+        int weight = calculateWeightFactor( periodType );
+
+        return Precision.round( (year1Value * ((double) (12 - weight) / 12)) + (year2Value * ((double) weight / 12)), 2 );
```
Comment (jimgrace) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/JdbcAnalyticsManager.java]: Scale factor 2 is too small in general. Each indicator has a "precision" that determines how many digits after the decimal place are returned. If you have handy access to that precision, you could use it here. But if you do not... :) Numerators and Denominators will eventually be trimmed to this precision before sending them to the frontend -- see for example in DefaultAnalytics Service line 531:
```
.addValue( AnalyticsUtils.getRoundedValue( dataSourceParams, indicator.getDecimals(), value.getDenominatorValue() ) )
```
When defining an indicator, the UI allows you to select a precision of 0,1,2,3,4,5, or `<No value>`. If an indicator does not have a precision, a value of 10 digits is used, as you can see if you search for `DECIMALS_NO_ROUNDING` as defined in AnalyticsUtils. So if you round to 10 digits here I think that will handle all cases, and the value may be further rounded before it is sent to the frontend. I suggest a symbolic constant -- like you might make `DECIMALS_NO_ROUNDING` public and use it here.

Comment (luciano-fiandesio) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/JdbcAnalyticsManager.java]: Ok, I went the simplest path and simply made `DECIMALS_NO_ROUNDING` public.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/FinancialAprilPeriodType.java [hunk 4]
```diff
@@ -79,4 +79,9 @@ public String getIso8601Duration()
         return ISO8601_DURATION;
     }
 
+    @Override
+    public boolean isFinancialYear() {
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/FinancialAprilPeriodType.java]: Codestyle

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/FinancialJulyPeriodType.java [hunk 5]
```diff
@@ -79,4 +79,9 @@ public String getIso8601Duration()
         return ISO8601_DURATION;
     }
 
+    @Override
+    public boolean isFinancialYear() {
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/FinancialJulyPeriodType.java]: Codestyle

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/FinancialNovemberPeriodType.java [hunk 6]
```diff
@@ -77,4 +77,9 @@ public String getIso8601Duration()
     {
         return ISO8601_DURATION;
     }
+
+    @Override
+    public boolean isFinancialYear() {
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/FinancialNovemberPeriodType.java]: Codestyle

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/PeriodType.java [hunk 7]
```diff
@@ -763,4 +763,8 @@ public String toString()
     {
         return "[" + getName() + "]";
     }
+
+    public boolean isFinancialYear() {
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/PeriodType.java]: Codestyle

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/FinancialNovemberPeriodType.java (hunk 6 above)
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/FinancialNovemberPeriodType.java]: I think these methods can be moved up to class `FinancialPeriodType`

**Pull request merged**